File: LeetCode/Medium/2684-maximum-number-of-moves-in-a-grid/2684-maximum-number-of-moves-in-a-grid.java
class Solution {
    public int maxMoves(int[][] grid) {
        int rowLen = grid.length;
        int colLen = grid[0].length;
        
        boolean[] prev = new boolean[rowLen];

        for (int i = 0; i < rowLen; ++i) prev[i] = true;
        
        for (int col = 1; col < colLen; ++col){
            boolean[] curr = new boolean[rowLen];

            for (int row = 0; row < rowLen; ++row){
                if (row > 0 && grid[row-1][col-1] < grid[row][col]) curr[row] |= prev[row-1];
                if (row < rowLen - 1 && grid[row+1][col-1] < grid[row][col]) curr[row] |= prev[row+1];
                if (grid[row][col-1] < grid[row][col]) curr[row] |= prev[row];
            }
            
            boolean isGood = false;
            for (int i = 0; i < rowLen; ++i)
                if (curr[i]){
                    isGood = true;
                    break;
                }
            
            if (!isGood) return col - 1;
            
            prev = curr;
        }

        return colLen - 1;
    }
}

File: LeetCode/Medium/2684-maximum-number-of-moves-in-a-grid/2684-maximum-number-of-moves-in-a-grid.py
class Solution:
    def maxMoves(self, grid: List[List[int]]) -> int:
        ret = 0
        row_len = len(grid)
        col_len = len(grid[0])
        
        prev = [True] * row_len

        for col in range(1, col_len):
            curr = [False] * row_len

            for row in range(row_len):
                if row and grid[row-1][col-1] < grid[row][col]:
                    curr[row] |= prev[row-1]
                if row < row_len - 1 and grid[row+1][col-1] < grid[row][col]:
                    curr[row] |= prev[row+1]
                if grid[row][col-1] < grid[row][col]:
                    curr[row] |= prev[row]

            if not sum(curr):
                return col - 1
            
            prev = curr

        return col_len - 1